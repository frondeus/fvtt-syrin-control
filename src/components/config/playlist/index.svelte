<script lang="ts">
	import Context from '@/services/context';
	import type { Soundsets } from '@/models/store';

	// Context
	const ctx = Context();

	// Params & State
	export let name: string;
	export let sorting: string;
	export let soundsetId: string;
	let soundsets = ctx.stores.soundsets;
	let soundsetName = '...';
	let style = '';
	let class_ = 'inner';

	const reactiveSoundsetName = async (soundsetId: string, soundsets: Soundsets) => {
		let soundset = soundsets[soundsetId];
		if (soundset === undefined) {
			return;
		}
		soundsetName = soundset.name;
		if (soundset.artworkUrl !== undefined) {
			style = `background-image: url('${soundset.artworkUrl}');`;
			class_ = 'inner inner-invert';
		}
	};

	$: reactiveSoundsetName(soundsetId, $soundsets);
</script>

<div class={class_} {style} data-test="syrin-playlist-config">
	<div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('PLAYLIST.Name')}</label>
			<input
				type="text"
				name="name"
				data-test="syrin-playlist-name"
				placeholder={ctx.game.localizeCore('PLAYLIST.Name')}
				bind:value={name}
			/>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localize('config.soundset')}</label>
			<input
				type="text"
				disabled
				data-test="syrin-soundset-name"
				value={soundsetName}
				title={soundsetName}
			/>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('PLAYLIST.SortMode')}</label>
			<select name="sorting" data-test="syrin-sort-mode" bind:value={sorting}>
				<option value="a">{ctx.game.localizeCore('PLAYLIST.SortAlphabetical')}</option>
				<option value="m">{ctx.game.localizeCore('PLAYLIST.SortManual')}</option>
			</select>
		</div>
	</div>
	<div>
		<div data-test="syrin-controlled">
			{ctx.game.localize('config.controlled', { name: ctx.game.localize('config.playlist') })}
		</div>
		<button type="submit">
			<i class="far fa-save" />
			{ctx.game.localizeCore('PLAYLIST.Update')}
		</button>
	</div>
</div>

<style>
	.inner-invert {
		color: var(--color-text-light-highlight);
		background-size: cover;
	}
	.inner {
		padding: 8px;
		display: flex;
		flex-direction: column;
		justify-content: space-between;
		height: 100%;
	}
	.inner button,
	.inner input,
	.inner select {
		background: url(../ui/parchment.jpg) repeat;
	}
</style>
