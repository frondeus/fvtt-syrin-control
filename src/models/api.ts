// What we get from Syrinscape
export interface ApiSoundset {
	id: number;
	uuid: string;
	name: string;
	full_name: string;
	artwork_url: string;
}

export interface ApiMood {
	pk: number;
	name: string;
	elements: ApiMoodElement[];
	soundset_uuid: string;
}

export interface ApiMoodElement {
	element: string; // Element url... last segment is pk
}

export interface ApiElement {
	name: string;
	pk: number;
	icon?: string;
	element_type: 'sfx' | 'oneshot';
}

export interface ApiStatus {
	global?: {
		soundset_uuid?: string;
	};
}
