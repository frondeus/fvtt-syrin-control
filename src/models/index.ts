export * from './api';
export * from './store';
