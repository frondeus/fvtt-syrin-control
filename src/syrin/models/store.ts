export interface Soundset {
	id: string;
	pid: number;
	name: string;
	moods: Moods;
	elements: Elements;
}

// export interface NextMood {
// 	moodId?: number;
// 	soundsetId?: number;
// }

export interface CurrentlyPlaying {
	mood: Mood;
	soundset: Soundset;
}

export interface Mood {
	id: number;
	name: string;
	soundset?: string;
	elementsIds: number[];
}

export interface Moods {
	[index: number]: Mood;
}

export interface Soundsets {
	[index: string]: Soundset;
}

export interface PlaylistEntry {
	mood: Mood;
	soundset: Soundset;
}

export interface Playlist {
	entries: PlaylistEntry[];
}

export interface Element {
	id: number;
	name: string;
	icon: string;
	type: "oneshot" | "sfx" | "music";
}

export interface Module {
	elementsApp: any;
}

export type Elements = Element[];

export interface GlobalElementsTab {
	kind: 'global';
}

export interface SoundsetElementsTab {
	kind: 'soundset';
	soundset: Soundset;
}

export type ElementsTab = SoundsetElementsTab | GlobalElementsTab;
export type ElementsTabs = ElementsTab[];

export interface PlaylistItem {
	isPlaying: boolean;
	mood: Mood;
	soundset: Soundset;
}

export interface AmbientSounds {
	[index: string]: AmbientSound
}

export type AmbientSound = MoodAmbientSound | ElementAmbientSound;

export interface MoodAmbientSound {
	kind: 'mood';
	volume: number;
	moodId: number;
}

export interface ElementAmbientSound {
	kind: 'element';
	volume: number;
	elementId: number;
}
