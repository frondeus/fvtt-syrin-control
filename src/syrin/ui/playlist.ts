import PlaylistComponent from '@/components/Playlist.svelte';
import PlayerVolumeComponent from '@/components/PlayerVolume.svelte';
import GMVolumeComponent from '@/components/GMVolume.svelte';
import { Context } from '@/services/context';
import { openMacroManager } from '@/ui/macromanager';

export async function onPlaylistTab(ctx: Context, $tab: JQuery<Element>) {
	let volumeTarget = $tab.find('.playlist-sounds');
	
	new PlayerVolumeComponent({
		target: volumeTarget.get(0)!,
		context: ctx.map()
	});

	if (!ctx.game.isGM()) {
		return;
	}
	
	new GMVolumeComponent({
		target: volumeTarget.get(0)!,
		context: ctx.map()
	});

	let target = $tab.find('.directory-list');

	// ctx.utils.trace('On Playlist Tab', { $tab });
	

	new PlaylistComponent({
		target: target.get(0)!,
		context: ctx.map()
	});
	
	const buttonsTarget = $tab.find('.action-buttons');
	const importButton = $(`<button class="import-syrinscape">\
		<i class="fas fa-file-import"></i>\
		${ctx.game.localize("playlist.importSoundsets")}\
		</button>\
		`);
	
	importButton.on('click', () => {
		openMacroManager(ctx);
	});
	
	buttonsTarget.append(importButton);
}
