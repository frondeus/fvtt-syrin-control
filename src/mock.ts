import { RawApi } from '@/services/raw';
import { FVTTGame } from '@/services/game';
import { storeDefaults, Stores } from '@/services/stores';
import { Utils } from './services/utils';
import { Api } from './services/api';
import { Syrin } from './services/syrin';
import { Context } from './services/context';

export function mockedApi(): RawApi {
	return {
		stopMood: jest.fn(),
		playMood: jest.fn(),
		playElement: jest.fn(),
		getMoods: jest.fn(async (_) => []),
		getElements: jest.fn(async (_) => []),
		getSoundsets: jest.fn(async () => []),
		getGlobalElements: jest.fn(async () => []),
		onInit: jest.fn(),
		getState: jest.fn(),
		changePlayerVolume: jest.fn(),
		changeMoodVolume: jest.fn(),
		changeOneShotVolume: jest.fn(),
		playerJoined: jest.fn(),
		stopElement: jest.fn(),
		getMood: jest.fn()
	};
}

type StoreDefaultsKey = keyof typeof storeDefaults;
export function mockedGame(): FVTTGame {
	return {
		getSetting: jest.fn((name) => {
			if (Object.hasOwn(storeDefaults as object, name)) {
				return storeDefaults[name as StoreDefaultsKey] as any;
			}
			console.error('get setting', { name });
			return null!;
		}),
		setSetting: jest.fn(),
		setGlobal: jest.fn(),
		localize: jest.fn((key) => key),
		registerSetting: jest.fn(),
		notifyInfo: jest.fn(),
		notifyError: jest.fn(),
		callHookAll: jest.fn(),
		getActiveScene: jest.fn(),
		isGM: jest.fn(),
		setSettingToDefault: jest.fn(),
		socket: {
			register: jest.fn(),
			executeAsGM: jest.fn(),
			executeAsUser: jest.fn(),
			executeForAllGMs: jest.fn(),
			executeForEveryone: jest.fn(),
			executeForOtherGMs: jest.fn(),
			executeForOthers: jest.fn(),
			executeForUsers: jest.fn()
		},
		isReady: jest.fn(() => true),
		userId: jest.fn(),
		hasActiveModule: jest.fn(),
		localizeCore: jest.fn(),
		getAudioContext: jest.fn(),
		createMoodMacro: jest.fn(),
		createElementMacro: jest.fn(),
		createPlaylist: jest.fn(),
		createPlaylistSound: jest.fn(),
		createPlaylistMoodSound: jest.fn(),
		getPlaylists: jest.fn(),
		getAmbientSounds: jest.fn(),
		getPlayerName: jest.fn(),
		createDialog: jest.fn()
	};
}

export function mocked(game: FVTTGame = mockedGame(), raw: RawApi = mockedApi()): Context {
	const utils = new Utils(game);
	const api = new Api(raw);
	const stores = new Stores(game, api);
	const syrin = new Syrin(game, utils, api, stores);
	return new Context(game, utils, api, stores, syrin);
}
