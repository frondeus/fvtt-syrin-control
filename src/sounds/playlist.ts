import { Context } from '@/services/context';
import { Unsubscriber } from 'svelte/store';

export interface PlaylistProvider {
	update(playing: boolean): void;
	id(): null | string;
}

export interface SyrinPlaylistFlags {
	soundset: string;
}

export class Playlist {
	ctx: Context;
	flags: SyrinPlaylistFlags;
	unsubscriber?: Unsubscriber;
	provider: PlaylistProvider;

	constructor(data: any, ctx: Context, provider: PlaylistProvider) {
		this.ctx = ctx;
		this.flags = data.flags.syrinscape;
		this.provider = provider;

		this.handleSubscribtion();
	}

	handleSubscribtion() {
		this.unsubscriber = this.ctx.stores.currentlyPlaying.subscribe((playing) => {
			const soundset = playing?.soundset;
			if (this.provider.id() !== null) {
				const playing = soundset?.id === this.flags.soundset;
				if (this.ctx.game.isGM()) {
					this.provider.update(playing);
				}
			}
		});
	}

	stopAll() {
		this.ctx.syrin.stopAll();
	}

	unsubscribe() {
		this.unsubscriber?.call([]);
	}
}
