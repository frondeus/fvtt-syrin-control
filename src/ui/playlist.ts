import PlaylistComponent from '@/components/Playlist.svelte';
import PlayerVolumeComponent from '@/components/PlayerVolume.svelte';
import GMVolumeComponent from '@/components/GMVolume.svelte';
import { Context } from '@/services/context';
import { openImporter } from '@/ui/importer';
import { SyrinComponent } from '@/services/syrin';

export async function onPlaylistTab(ctx: Context, $tab: JQuery<Element>) {
	let volumeTarget = $tab.find('.playlist-sounds');

	ctx.syrin.renderComponent(
		ctx,
		SyrinComponent.PlayerVolume,
		PlayerVolumeComponent,
		volumeTarget.get(0)!
	);

	if (!ctx.game.isGM()) {
		return;
	}

	ctx.syrin.renderComponent(ctx, SyrinComponent.GMVolume, GMVolumeComponent, volumeTarget.get(0)!);

	let target = $tab.find('.directory-list');

	// ctx.utils.trace('On Playlist Tab', { $tab });

	ctx.syrin.renderComponent(ctx, SyrinComponent.Playlist, PlaylistComponent, target.get(0)!);

	const buttonsTarget = $tab.find('.action-buttons');
	const importButton = $(`<button class="import-syrinscape" data-test="syrin-import-btn">\
		<i class="fas fa-file-import"></i>\
		${ctx.game.localize('playlist.importSoundsets')}\
		</button>\
		`);

	importButton.on('click', () => {
		openImporter(ctx);
	});

	buttonsTarget.append(importButton);
}
