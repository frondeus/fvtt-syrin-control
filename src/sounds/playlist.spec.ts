import { Mocked, mocked } from '@/mock';
import { ApiElement } from '@/models';
import { Playlist, SyrinPlaylistFlags } from './playlist';

describe('Ambient sound controller', () => {
	let mock: Mocked;
	let flags: SyrinPlaylistFlags;

	beforeEach(async () => {
		mock = mocked();
		mock.game.isGM = jest.fn(() => true);
		mock.game.userId = jest.fn(() => 'frondeus');
		mock.raw.getState = jest.fn(() => 'active');

		flags = {
			soundset: '1234'
		};

		mock.raw.getMood = jest.fn(async (id) => {
			if (id === 4321) {
				return await import('@fixtures/raw-mood-4321.json');
			}
			return undefined;
		});
		mock.ctx.stores.soundsets.set({
			'1234': (await import('@fixtures/soundset-1234.json')).default
		});
		mock.raw.getElements = jest.fn(async (id) => {
			if (id === '1234') {
				return (await import('@fixtures/raw-elements-1234.json')).default as ApiElement[];
			}
			return [];
		});
	});

	it('calls unsubscriber', () => {
		const provider = {
			id: () => '1',
			update: jest.fn((_playing: boolean) => {}),
			ctx: () => mock.ctx
		};
		const sound = new Playlist({ flags: { syrinscape: flags } }, provider);
		const spy = jest.spyOn(sound, 'unsubscriber');

		sound.unsubscribe();

		expect(spy).toBeCalledTimes(1);
	});

	it('stops all', () => {
		const provider = {
			id: () => '1',
			update: jest.fn((_playing: boolean) => {}),
			ctx: () => mock.ctx
		};
		const sound = new Playlist({ flags: { syrinscape: flags } }, provider);
		const spy = jest.spyOn(mock.ctx.syrin, 'stopAll');

		sound.stopAll();

		expect(spy).toBeCalledTimes(1);
	});

	describe('subscribtion', () => {
		it('changes playing state when currentlyPlaying store is updated', async () => {
			let newPlaying = false;
			const provider = {
				id: () => '1',
				update: jest.fn((playing: boolean) => {
					newPlaying = playing;
				}),
				ctx: () => mock.ctx
			};

			expect(newPlaying).toBe(false);

			const sound = new Playlist({ flags: { syrinscape: flags } }, provider);

			mock.ctx.stores.nextPlaylistMood.set(4321);

			await new Promise((resolve) =>
				mock.ctx.stores.currentlyPlaying.subscribe((playing) => {
					if (playing !== undefined) resolve(playing);
				})
			);

			expect(newPlaying).toBe(true);

			sound.unsubscribe();
		});
	});

	describe('bugs', () => {
		it('can be flattened', () => {
			const provider = {
				id: () => '1',
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new Playlist({ flags: { syrinscape: flags } }, provider);

			const flattened = foundry.utils.flattenObject(sound);

			expect(flattened).toMatchSnapshot();
		});
	});
});
