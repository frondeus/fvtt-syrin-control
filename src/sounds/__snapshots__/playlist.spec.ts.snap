// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ambient sound controller bugs can be flattened 1`] = `
{
  "flags.soundset": "1234",
  "provider.ctx": [Function],
  "provider.id": [Function],
  "provider.update": [MockFunction] {
    "calls": [
      [
        false,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "unsubscriber": [Function],
}
`;
